src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -2974,3 +2978,33 @@ EnsureShardAllowedOnNodeUDF(Oid functionOid)
 	}
 	ReleaseSysCache(proctup);
 }
+
+
+/*
+ * EnsureNoActiveShardPlacementsOnDisabledNodes raises an error if
+ * there are active shards on disabled worker nodes.
+ *
+ * This is used to disallow running rebalancing or replication operations on shards
+ * when some of the active shards are in diabled worker nodes
+ */
+static void
+EnsureNoActiveShardPlacementsOnDisabledNodes(List *shardPlacements)
+{
+	List *disabledNodes = DisabledNodeList();
+	WorkerNode *workerNode = NULL;
+	ShardPlacement *shardPlacement = NULL;
+
+	foreach_ptr(shardPlacement, shardPlacements)
+	{
+		foreach_ptr(workerNode, disabledNodes)
+		{
+			if (shardPlacement->shardState == SHARD_STATE_ACTIVE &&
+				shardPlacement->nodeId == workerNode->nodeId)
+			{
+				ereport(ERROR, (errmsg(
```

Comment:
We currently define active placements to be placements on active nodes. So, by definition, we should not  `ActiveShardPlacementsOnDisabledNodes`. 

It still not very clear to me why changing `ActivePlacementsHash` to consider this new definition does OK? We might need something additional, but it seems like that's a must have?

To me, without fixing that, we are still going to be vulnerable to such weird edge cases.

Reply:
In that case, if this is run on the coordinator:
```pgsql
SET citus.shard_replication_factor TO 2;
CREATE TABLE test(a int);
SELECT citus_disable_node('localhost', :worker2_id);
SELECT * FROM pg_dist_shard_placement;
```
All the shards will be reported, and they all will be labelled active. The active shards on the disabled node conflict with our definition of active, so we might want to consider to change the shard placement statuses of shards in disabled nodes.

Reply:
Ack., that is confusing. But unfortunately, currently the `shardstate` on `pg_dist_shard_placement ` *does not fully reflect* the shard's state. Maybe we should find a way to show it properly.


As we discussed in person, the new definition of `ActiveShardPlacementList()` does the following such that it should be on an active node:
```C
		if (shardPlacement->shardState == SHARD_STATE_ACTIVE &&
			workerNode->isActive)
		{
			activePlacementList = lappend(activePlacementList, shardPlacement);
		}
```


And, in simple words, the reason can be explained as follows:
(a) Citus never marks any placement `SHARD_STATE_INACTIVE` anymore. All modifications happens via 2PC. Because, on MX, it is non-trivial to propagate metadata changes from any node. (e.g., modifications can happen from any node), we preferred this approach. 
(b) Due to possibility of multiple node failures at the same time, we prefer to let a background daemon to sync the node metadata changes after citus_disable_node
(c) Once a node is activated, we make sure that the "disabled" placements are cleared properly: https://github.com/citusdata/citus/blob/d0aa450d7bc8bc4cebf6cf70ce15b07bbc65a780/src/backend/distributed/metadata/node_metadata.c#L1077-L1091   We do this as `activateNode` because it is much more likely that users activate nodes when all the cluster is up and running (compared to disable node, which happens to shield against failures) 


The more discussion on the PRs landed for this topic is here: https://github.com/citusdata/citus/issues/1033#issuecomment-940806179